### ISSUE TITLE
PMIx with tool support is removing everything in /tmp

### ISSUE THREAD

**Issue by dgloe-hpe** [openpmix/openpmix#3476]
### What version of the PMIx Reference Library are you using? 
5.0.3

### Describe how PMIx was installed 
From an RPM.

### Please describe the system on which you are running

* Operating system/version: SLES 15 SP5
* Computer hardware: Virtual Machine
* Network type: N/A (single system)

-----------------------------

## Details of the problem

I'm testing out adding tool support to our PMIx server by setting PMIX_SERVER_TOOL_SUPPORT to true and implementing the tool_connected handler. After doing this, I see `/tmp/pmix.<hostname>.tool.<pid>` and `/tmp/pmix.<hostname>.tool.<namespace>` files whenever an application is running, which I expected.

However, it looks like PMIx is removing everything from /tmp during cleanup, not just the rendezvous files.

Looking at the code, this appears to be due to `pmix_base_write_rndz_file` setting `*created = true`, even if the directory already exists.

I confirmed that pmix was the culprit using strace:
```
28661 write(2, "[sles15sp5:28661] listen_thread:"..., 42) = 42
28661 unlink("/tmp/pmix.sles15sp5.tool.pals.pmix.85d4af96-00ae-459d-a57c-f7244ccee401.(null)") = 0
28661 unlink("/tmp/pmix.sles15sp5.tool.28661") = 0
28661 openat(AT_FDCWD, "/tmp", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = 21
28661 fstat(21, {st_mode=S_IFDIR|0755, st_size=86016, ...}) = 0
28661 getdents64(21, 0x559cd136af10 /* 5 entries */, 32768) = 144
28661 unlink("/tmp/tmpzdyve9mv")        = -1 EISDIR (Is a directory)
28661 rmdir("/tmp/tmpzdyve9mv")         = -1 ENOTEMPTY (Directory not empty)
28661 openat(AT_FDCWD, "/tmp/tmpzdyve9mv", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = 22
28661 fstat(22, {st_mode=S_IFDIR|0700, st_size=4096, ...}) = 0
28661 getdents64(22, 0x559cd1349350 /* 3 entries */, 32768) = 104
28661 unlink("/tmp/tmpzdyve9mv/85d4af96-00ae-459d-a57c-f7244ccee401") = -1 EISDIR (Is a directory)
28661 rmdir("/tmp/tmpzdyve9mv/85d4af96-00ae-459d-a57c-f7244ccee401") = -1 ENOTEMPTY (Directory not empty)
...
28661 unlink("/tmp/hwloc.sm")           = 0
28661 getdents64(21, 0x559cd136af10 /* 0 entries */, 32768) = 0
28661 close(21)                         = 0
28661 write(2, "[sles15sp5:28661] mca: base: clo"..., 63) = 63
28661 write(2, "[sles15sp5:28661] mca: base: clo"..., 63) = 63
28661 write(2, "[sles15sp5:28661] mca: base: clo"..., 61) = 61
```

**Comment by rhc54** [comment]
When you call `PMIx_server_init`, do you pass down the upper-level session directory location for the server to use (via `PMIX_SERVER_TMPDIR`)? This will isolate the rendezvous files, shmem backing files, etc. Otherwise, there isn't any good way for PMIx to know what it created and what therefore needs to be removed.

I can explore alternative approaches - e.g., trying to track every individual file that gets created - but it gets a little tricky as some elements create directories (e.g., to have a session directory for each job).

**Comment by rhc54** [comment]
Hmmm...let me retract those words. PMIx does indeed track what it creates and should only remove those files. However, I do see that PRRTE mistakenly empties the top-level directory when told to operate as the "system-level" server. Let me look into this a bit. Not immediately certain what is being removed by PRRTE vs PMIx (both use the same functions to do the removal, so strace won't help there).

> Looking at the code, this appears to be due to pmix_base_write_rndz_file setting *created = true, even if the directory already exists.

I agree that would be the likely problem - appreciate the pointer! Will get back to you shortly.

**Comment by rhc54** [comment]
Well, I dug around a bit and found the root cause of the problem was a change made to the code that creates the directory tree where the various rendezvous files are to be stored. Grok why the change was made - unfortunately, neither the author nor I fully identified some of the consequences. 

The changes were valid, so I needed to modify things to compensate for those changes. I have something now that seems to do the job, but want to test it a bit more to ensure it doesn't cause unintended consequences of its own.

**Comment by rhc54** [comment]
@dgloe-hpe Would you be able/willing to try https://github.com/openpmix/openpmix/pull/3478 (ignore the yaml file - just added a PRRTE test to ensure this didn't break anything there)? I'm not sure, but think it should apply cleanly to the v5.x code you were using. If not, I'll be creating a port anyway.

**Comment by dgloe-hpe** [comment]
No, we don't set `PMIX_SERVER_TMPDIR`.
I'll try out the patch shortly, just got back from break today.

**Comment by dgloe-hpe** [comment]
That patch did resolve the problem for me. Do you plan on making a v5.0.x release including this fix?

**Comment by rhc54** [comment]
Yeah - should be out next week, I think. Thanks for bringing it my attention!

**Comment by rhc54** [comment]
Released in PMIx v5.0.6 - thanks again!
